### ISSUE TITLE
Student Developer Rights and Ownership of the Production 

### ISSUE THREAD

**Issue by thirdeyeclub** [Lambda-School-Labs/labs-engineering-standards#47]
I believe that we need to change section GH 101 needs to be changed so that BOTH Lambda Staff AND lambda students are owners of the organization.

My argument for this change will be broken into 3 parts. 

1. Responsibility and Reaction Time
2. Ethical Responsibility 
3.  Deconstruction of Counter Arguments and Issues

1 Responsibility and Reaction Time
Labs is fitted to be a real world analog to working on a team that is building a product. Reducing the Students, the primary role to that of only a member is taking away repeatability that would otherwise rest on their shoulders thus diminishing the experience of working as a full fledged team member. This separation of student from project ownership devalues their role within the organization, further more it causes technical challenges that would other wise solve themselves.  There is no benefit to having the student as only a member, other than creating a needless hierarchy which will get in the way of production. This is not to say to remove the roles of Team Lead or Section Lead as socially and within the context of the school they are needed. This IS however a call to equalize the responsibilities of the specific Labs project. The benefits of having students as direct owners is as follows. A simpler contribution flow that empowers the Students to review and merge each others code. Allow them to make changes to the reposiaity directly as needed without having to pass DMs along a social hierarchy which WILL cause production delays, delays in bug fixes, and delays in continued production post-labs. I have personally had issues in the past where I had to wait WEEKS for TLs to approve things, this is aggravating and does not need to happen. Also allowing students to directly control  POST-LABS production empowers their job search without causing headaches for LABS staff.

2 Ethical Responsibility 
Why do we use open source at Lambda? Because it allows us to contribute freely with multiple people and innovate. So who really owns the code if it's sitting freely on the internet for any one to use? The contributors of course, (depending on the license you chose to put on your repo). BUT the contributors are the ones who do create the code. The majority of which is written by the students. Therefore it is the ethical choose to give those who contribute to the code ownership of that code. This total ownership also promotes a more democratic (as long as it is maintained that there has to be reviews on pull requests) environment to project. Democracy requires cooperation and collaboration, thus increasing the students responsibilities and sense of pride/ownership of the code base therefore empowering them to make better more meaningful choices. To not allow students ownership of repositories that they contribute heavily to is exploitation. The students already own the means of production (IDEs, their computer, ect) they should have ownership over the product itself IN AN OFFICIAL SENSE being that of the ownership role. The fruits of the labor shall be owned by those who grow the fruit

3 Deconstruction of Counter Arguments and Issues
The first argument I will discuss is a technical one. Bad merges. This can consist of, merging untested code, broken code,  malicious code, anything production damaging or breaking. All of these become a very real possibility when students have ownership. But this possibility as I touched on in the first paragraph will be beneficial to the students experience beaccause they have to learn to self organize and abide by their own rules to create a working product. This will improve their experience. A double edge sword is no worry to those who are trained to swing it correctly.
Another disagreement or augment could be that the students will not have a predefined structured system to contribute. This is untrue as they have already been thought 3 build weeks at this point and are used to how to contribute, further more this puts the weight of self origination onto the student allowing them to disseminate/refactor/iterate over things that they disliked about  the conurbation systems they experienced in the past.

In conclusion I believe that this will empower students to make better decisions over all,  by letting them grow or fail by their own hands, learning from their failures. Less hierarchical authority from SLs and TLs (making staff have to merge code)  will allow faster development, bug fixes, innovation in the long run. Finally that it is the ethical responsibility as the students are the main contributors that they be the owners of what they create.

**Comment by berniedurfee** [comment]
Thanks for the feedback! I need to digest this one and probably run it past a few people. I'll circle back.

**Comment by berniedurfee** [comment]
Comments below...

> 1 Responsibility and Reaction Time
> Labs is fitted to be a real world analog to working on a team that is building a product. Reducing the Students, the primary role to that of only a member is taking away repeatability that would otherwise rest on their shoulders thus diminishing the experience of working as a full fledged team member. This separation of student from project ownership devalues their role within the organization, further more it causes technical challenges that would other wise solve themselves.

We limit students access to the overall org for reasons of separation of responsibilities, but mainly to practice the principle of least privilege. The goal is to give students the access they need to work on their products, but no more. Those boundaries keep the overall organization safer. Though, as with all security constraints, this will create overhead, but that's the balance we need to maintain.

Are there specific technical challenges this causes?

> There is no benefit to having the student as only a member, other than creating a needless hierarchy which will get in the way of production.

Unfortunately, security will always create inconveniences, such is the nature of security. As far as I know though, this particular situation isn't causing more than a occasional inconvenience. Is it?

> This is not to say to remove the roles of Team Lead or Section Lead as socially and within the context of the school they are needed. This IS however a call to equalize the responsibilities of the specific Labs project. The benefits of having students as direct owners is as follows. A simpler contribution flow that empowers the Students to review and merge each others code. Allow them to make changes to the reposiaity directly as needed without having to pass DMs along a social hierarchy which WILL cause production delays, delays in bug fixes, and delays in continued production post-labs. I have personally had issues in the past where I had to wait WEEKS for TLs to approve things, this is aggravating and does not need to happen. Also allowing students to directly control POST-LABS production empowers their job search without causing headaches for LABS staff.

Changes to repository configuration should be rare and are impactful, which is why we are careful to limit the number of individuals with those permissions. Once a repo is configured, there should be nothing that requires admin privileges to create a pipeline from commit to push to pr to merge to production.

Having said that, if there are delays of days or weeks for those occasional needs, that _is_ an issue. Feel free to reach out to me directly to let me know of any ongoing delays. There's likely a process issue that needs to be resolved.

As for post-Labs access, we are committed to keeping student codebases available as public repositories. These are currently all MIT licensed as well. This means that anyone, at any time, for any reason can access and fork the code for any purpose. Our goal is to protect and maintain the original repository for future teams to build-on, but that doesn't limit access or use of that codebase by anyone else.

> 2 Ethical Responsibility
> Why do we use open source at Lambda? Because it allows us to contribute freely with multiple people and innovate. So who really owns the code if it's sitting freely on the internet for any one to use? The contributors of course, (depending on the license you chose to put on your repo). BUT the contributors are the ones who do create the code. The majority of which is written by the students. Therefore it is the ethical choose to give those who contribute to the code ownership of that code. This total ownership also promotes a more democratic (as long as it is maintained that there has to be reviews on pull requests) environment to project. Democracy requires cooperation and collaboration, thus increasing the students responsibilities and sense of pride/ownership of the code base therefore empowering them to make better more meaningful choices. To not allow students ownership of repositories that they contribute heavily to is exploitation. The students already own the means of production (IDEs, their computer, ect) they should have ownership over the product itself IN AN OFFICIAL SENSE being that of the ownership role. The fruits of the labor shall be owned by those who grow the fruit

This is why we chose and maintain the use of the MIT license. We feel it upholds our commitment to our students that they own the code they write. Org level permissions don't reflect on access to or ownership of the code; it's the license that matters, which in our case, is MIT, which is very flexible and open.


> 3 Deconstruction of Counter Arguments and Issues
> The first argument I will discuss is a technical one. Bad merges. This can consist of, merging untested code, broken code, malicious code, anything production damaging or breaking. All of these become a very real possibility when students have ownership. But this possibility as I touched on in the first paragraph will be beneficial to the students experience beaccause they have to learn to self organize and abide by their own rules to create a working product. This will improve their experience. A double edge sword is no worry to those who are trained to swing it correctly.

I agree that students should be able to make mistakes. Mistakes are the best opportunities to learn. At the same time, we also need to ensure that our program can scale and we are always balancing creative freedom and our ability to keep the wheels turning. In fact, the constraints under which teams need to produce is a learning experience in and of itself. No organization allows total freedom of access and decision making; thus the hierarchical nature of business organizations.

> Another disagreement or augment could be that the students will not have a predefined structured system to contribute. This is untrue as they have already been thought 3 build weeks at this point and are used to how to contribute, further more this puts the weight of self origination onto the student allowing them to disseminate/refactor/iterate over things that they disliked about the conurbation systems they experienced in the past.

Again, it's all about balance and compromise. In Labs, we have a unique and rather challenging environment. There's only so much heterogeneity we can manage before the complexity becomes a real detriment to our ability to support learning. This is a constant source of internal discussion that pushes and pulls us in various directions. Self-organization and total creative freedom is one side of the spectrum; total command and control on the other; I feel that we strike a good balance and certainly lean toward allowing creativity. 

> In conclusion I believe that this will empower students to make better decisions over all, by letting them grow or fail by their own hands, learning from their failures. Less hierarchical authority from SLs and TLs (making staff have to merge code) will allow faster development, bug fixes, innovation in the long run. Finally that it is the ethical responsibility as the students are the main contributors that they be the owners of what they create.

In almost all cases, students themselves do manage merging code into master, which will be pushed to production, without needing even approval or action from team or section leads. We require peer reviews because we want students to product quality software, of which peer reviews are a requirement. Peer reviews also provide an opportunity for students to learn from each other.

There doesn't appear to be any compelling and specific reason to change our permission structure at this time. When it comes to security, we need to ensure that decisions are made very carefully and that broad changes are only made when there is a significant and recurring impact. Public repositories and MIT licensing ensures that students have as much right to the code as anyone else. The 'Owner' role in GitHub only controls permissions for access to repositories, which is something that needs to be controlled to maintain the principle of least privilege.

Keeping Labs moving and providing an awesome experiential learning experience is a balancing act. I appreciate your position and frankly, if security and our ability to scale the program weren't limiting factors, I would agree that more freedom of creativity and ownership often leads to better outcomes. But such is the balance we need to keep. Thank you _very_ much for your input!
